jena-core/src/main/java/org/apache/jena/assembler/ModelExpansion.java

```diff
@@ -179,9 +179,8 @@ protected static void addDomainTypes( Model result, Model schema )
             Statement s = it.nextStatement();
             Property property = s.getSubject().as( Property.class );
             RDFNode type = s.getObject();
-            for (StmtIterator x = result.listStatements( ANY, property, ANY ); x.hasNext();)
+                for (Statement t : result.listStatements( ANY, property, ANY ).toList())
```

Comment:
The next line is :  "result.add( t.getSubject(), RDF.type, type );" --> so it was mutating the source of the iterator.
That worked with GraphMem because its iterator was counting backwards. But I guess it would also have caused an error if the array needed to resize due to the .add() operation.

Reply:
The contract only needs to be the usual Java contract on iterators. Just because GraphMem happened to work in some other ways isn't, to me, much of an issue to worry about. Ditto remove on iterators; a lot of Jena iterators do not support remove so good user code to work with any graph can't rely on remove.

Circumstantial  evidence - java.util.Iterator.remove is now a default method "UnsupportedOperationException".

Reply:
That is a relief to me but some code, especially tests, in the jena project uses Iterator.remove. 
I needed to support it to run all tests of the project.